### ISSUE TITLE
there is no aggregation function

### ISSUE THREAD

**Issue by Nanyan** [kwantam/bls_sigs_ref#1]
The current implementation does not seem to implement the aggregation of signatures. Will you plans to implement those functions?

**Comment by burdges** [comment]
Actually this matters because BLS signatures should never be used without aggregation.  In particular, BLS should never be used when you only want a VRF, not an aggregate/threshold VRF. 

There is a question of whether you use delinearization or proof-of-possession.  I initially made the mistake of believing proof-of-possession to be more efficient when writing https://github.com/w3f/bls but actually this is false even for consensus protocols.  

In https://crypto.stanford.edu/~dabo/pubs/papers/BLSmultisig.html Dan Boneh describes doing delinearization for account based schemes, which works, but sounds extremely slow.

Instead, you should adapt BLS only to consensus protocols by adding a second KeyPrep phase after KeyGen and KeySetPrep phase before Verify, which exponentiate the secret and public keys, respectively, based on the key set.  After this modification, you'll observe the prep phases run much faster than a proof-of-possession check, but otherwise the system behaves like a proof-of-possession system.

In other words, the aggregation function has two parts (a) transform the signer's secret key based on the signer set and (b) transform an entire signer set into a signer set ready for verification.  There is no analog on (a) in a proof-of-possession system, but realistically signers should always know the signer set.  And (b) exactly matches the proof-of-possession check, except it runs far faster, and may prove more amenable to type system level protections against miss-use in modern languages like Rust.

**Comment by kwantam** [comment]
There are now aggregation functions in the Python and Rust code that are consistent with the current BLS signatures draft, so I think it is safe to close this issue. Does that seem correct to you, @Nanyan?

@burdges, what you're suggesting seems to be a way of doing aggregation that isn't the same as any of the methods in the BLS draft. I'll have to think about your comment more carefully, but perhaps if this is a change you'd like to see in the BLS standard, you can open an issue in [that repo](https://github.com/cfrg/draft-irtf-cfrg-bls-signature) describing your proposal, and we can discuss further?

**Comment by burdges** [comment]
I never wrote down the full trade offs between my proposed method and the more common proof-of-possession method, but yeah maybe that's a reasonable venue.  

I think proofs-of-possession turn out faster whenever you cache verifications and have low but non-zero signer turnover.  If you've no caching or high turnover, then my interpretation of Dan Boneh's scheme should turn out faster.

**Comment by Nanyan** [comment]
@kwantam Thanks for your great work, we will consider to evaluate it, maybe later. It's OK to close this issue base on your consideration.

Thanks again.
